packages/compass-crud/src/stores/crud-store.js

```diff
@@ -513,38 +518,58 @@ const configureStore = (options = {}) => {
       track('Document Updated', { mode: this.modeForTelemetry() });
       try {
         doc.emit('update-start');
-        const object = doc.generateObject();
-        const query = doc.getQueryForOriginalKeysAndValuesForSpecifiedKeys([
-          '_id',
-          ...Object.keys(this.state.shardKeys || {}),
-        ]);
-        debug('Performing findOneAndReplace', { query, object });
 
         if (!(await this._verifyUpdateAllowed(this.state.ns, doc))) {
           // _verifyUpdateAllowed emitted update-error
           return;
         }
 
+        const object = doc.generateObject();
+        const queryKeyInclusionOptions = {
+          alwaysIncludeKeys: [
+            '_id',
+            // '.' in shard keys means nested doc
+            ...Object.keys(this.state.shardKeys || {}).map((key) =>
+              key.split('.')
+            ),
+          ],
+        };
+
         if (
           this.dataService.getCSFLEMode &&
-          this.dataService.getCSFLEMode() === 'enabled' &&
-          object.__safeContent__ &&
-          isEqual(
-            object.__safeContent__,
-            doc.generateOriginalObject().__safeContent__
-          ) &&
-          (
-            await this.dataService
-              .getCSFLECollectionTracker()
-              .knownSchemaForCollection(this.state.ns)
-          ).hasSchema
+          this.dataService.getCSFLEMode() === 'enabled'
         ) {
-          // SERVER-66662 blocks writes of __safeContent__ for queryable-encryption-enabled
-          // collections. We remove it unless it was edited, in which case we assume that the
-          // user really knows what they are doing.
-          delete object.__safeContent__;
+          const knownSchemaForCollection = await this.dataService
+            .getCSFLECollectionTracker()
+            .knownSchemaForCollection(this.state.ns);
+
+          // The find/query portion will typically exclude encrypted fields,
+          // because those cannot be queried to make sure that the original
+          // value matches the current one; however, if we know that the
+          // field is equality-searchable, we can (and should) still include it.
+          queryKeyInclusionOptions.includableEncryptedKeys =
+            knownSchemaForCollection.encryptedFields.equalityQueryableEncryptedFields;
+
+          if (
+            object.__safeContent__ &&
+            isEqual(
+              object.__safeContent__,
+              doc.generateOriginalObject().__safeContent__
+            ) &&
+            knownSchemaForCollection.hasSchema
+          ) {
+            // SERVER-66662 blocks writes of __safeContent__ for queryable-encryption-enabled
+            // collections. We remove it unless it was edited, in which case we assume that the
+            // user really knows what they are doing.
+            delete object.__safeContent__;
+          }
         }
 
+        const query = doc.getQueryForOriginalKeysAndValuesForSpecifiedKeys(
```

Comment:
I sometimes think about whether or not I can use long function names, but omg this is such a beautiful and self-explanatory name! Won't hesitate in the future :)